browser/extensions/extensibles/test/browser/browser_privatetab.js

```diff
@@ -27,18 +41,152 @@ add_task(async function testContainer() {
-  let tab = gBrowser.selectedTab;
+  let tab = await BrowserTestUtils.openNewForegroundTab(gBrowser);
   is(tab.userContextId, 0, "Tab is not private");
-  await openTabContextMenu(tab);
-  let openPrivate = document.getElementById("toggleTabPrivateState");
-  openPrivate.click();
-  tab = gBrowser.selectedTab;
+  // Toggling private opens a duplicate tab in new container
+  tab = await togglePrivate(tab);
   is(
     tab.userContextId,
     window.PrivateTab.container.userContextId,
     "Tab is private"
   );
+  // Cleanup
+  BrowserTestUtils.removeTab(tab);
+});
+
+// Test autofill not stored from private tab
+add_task(async function testAutofillNotStored() {
+  const URL =
+    "http://mochi.test:8888/browser/browser/components/" +
+    "sessionstore/test/browser_formdata_sample.html";
+
+  const OUTER_VALUE = "browser_formdata_" + Math.random();
+  const INNER_VALUE = "browser_formdata_" + Math.random();
+
+  // Creates a tab, loads a page with some form fields,
+  // modifies their values and closes the tab.
+  async function createAndRemoveTab() {
+    // Create a new tab.
+    let tab = await BrowserTestUtils.openNewForegroundTab(gBrowser, URL);
+    // Toggle to a private tab
+    tab = await togglePrivate(tab, true);
+    let browser = tab.linkedBrowser;
+    await promiseBrowserLoaded(browser);
+
+    // Modify form data.
+    await setPropertyOfFormField(browser, "#txt", "value", OUTER_VALUE);
+    await setPropertyOfFormField(
+      browser.browsingContext.children[0],
+      "#txt",
+      "value",
+      INNER_VALUE
+    );
+
+    // Remove the tab.
+    await promiseRemoveTabAndSessionState(tab);
+  }
+
+  await createAndRemoveTab();
+  let [
+    {
+      state: { formdata },
+    },
+  ] = JSON.parse(SessionStore.getClosedTabData(window));
+  ok(!formdata, "Form data not stored in private tab");
+});
+
+// Test tab history not stored from private tab
+add_task(async function testTabHistoryNotStored() {
+  // Open a new tab
+  let tab = await BrowserTestUtils.openNewForegroundTab(gBrowser, URI1);
+  // Make it private
+  tab = await togglePrivate(tab, true);
+  let browser = tab.linkedBrowser;
+  await promiseBrowserLoaded(browser);
+  // Open a new URL
+  BrowserTestUtils.loadURI(browser, URI2);
+  await promiseBrowserLoaded(browser, false, URI2);
+  // Remove tab to save state
+  BrowserTestUtils.removeTab(tab);
+  // Verify only non-private data stored
+  let closedTabData = JSON.parse(SessionStore.getClosedTabData(window)).filter(
+    data => {
+      return (
+        data.state.entries[0].url == URI1 || data.state.entries[0].url == URI2
+      );
+    }
+  );
+  let privateData = closedTabData.filter(data => {
+    return data.state.isPrivate == true;
+  });
+  is(closedTabData.length, 1, "Tab data not stored in private mode");
+  is(privateData.length, 0, "Tab data not stored in private mode");
+});
+
+// Test private->non-private tab is restored on restart
+add_task(async function testRestoreConvertedPrivateTab() {});
```

Comment:
TBD or something?

Reply:
Yeah this is something that we need to be aware of but I haven't yet found a nice way of testing it. It's going to be quite an involved test as we need to create a new private tab 1) directly 2) by converting a non-private tab with the `togglePrivate()` method. After a restart neither of these tabs should be present, but need to have a nice way of triggering a restart and maintaining access to the restarted window object in the test. We also need to do the same thing of creating two private tabs and then converting them to non-private tabs. After a restart both of these should be present as non-private tabs. There was an issue raised a few weeks ago whereby creating a normal tab -> toggling to private -> toggling back to normal would cause the tab to be restored as a private tab.